hw/ip/otbn/util/shared/decode.py

```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
+import struct
+from typing import Dict, Tuple
+
+from shared.elf import read_elf
+from shared.insn_yaml import Insn, load_insns_yaml
+
+# Load the insns.yml file at module load time.
+try:
+    INSNS_FILE = load_insns_yaml()
+except RuntimeError as err:
+    sys.stderr.write('{}\n'.format(err))
+    sys.exit(1)
+
+
+class OTBNProgram:
+    def __init__(self, symbols: Dict[str, int], insns: Dict[int, int],
+                 data: Dict[int, int]):
+        self.symbols = symbols  # label -> PC
+        self.data = data  # addr -> data (32b word)
+
+        self.insns = {}
+        for pc in insns.keys():
+            opcode = insns[pc]
+            mnem = INSNS_FILE.mnem_for_word(opcode)
+            if mnem is None:
+                raise ValueError('No legal decoding for mnemonic: {}'.format(mnem))
+            insn = INSNS_FILE.mnemonic_to_insn.get(mnem)
+            enc_vals = insn.encoding.extract_operands(opcode)
+            op_vals = insn.enc_vals_to_op_vals(pc, enc_vals)
+            self.insns[pc] = (insn, op_vals)
+
+    def get_pc_at_symbol(self, symbol: str) -> int:
+        return self.symbols[symbol]
+
+    def get_data_at_addr(self, addr: int) -> int:
+        return self.data[addr]
+
+    def get_insn(self, pc: int) -> Insn:
+        return self.insns[pc][0]
+
+    def get_operands(self, pc: int) -> Dict[str, int]:
```

Comment:
Is it worth having a stronger type for this than a Dict?

Reply:
There's more type information in the instruction `Insn` instance, which has an `operands` attribute that gives you the type, name, etc of each operand for the instruction. Then this dict just matches names to values for this particular PC.